File: backend/app/schemas/request.py
from datetime import date
from pydantic import BaseModel, Field, ConfigDict
from app.models.request import PickupTimeSlot, RequestStatus


class RequestCreate(BaseModel):
    pickup_address: str = Field(max_length=255)
    pickup_postcode: str = Field(max_length=12)

    dropoff_address: str = Field(max_length=255)
    dropoff_postcode: str = Field(max_length=12)

    pickup_date: date
    pickup_time_slot: PickupTimeSlot

    helpers_needed: int = Field(ge=0)

    pickup_floor: int = Field(ge=0)
    pickup_has_lift: bool = False

    dropoff_floor: int = Field(ge=0)
    dropoff_has_lift: bool = False

    pickup_loading_minutes: int = Field(ge=0)
    dropoff_loading_minutes: int = Field(ge=0)


class RequestResponse(BaseModel):
    id: int
    user_id: int

    pickup_address: str
    pickup_postcode: str

    dropoff_address: str
    dropoff_postcode: str

    pickup_date: date
    pickup_time_slot: PickupTimeSlot

    helpers_needed: int

    pickup_floor: int
    pickup_has_lift: bool

    dropoff_floor: int
    dropoff_has_lift: bool

    pickup_loading_minutes: int
    dropoff_loading_minutes: int

    request_status: RequestStatus

    #deprecated
    #class Config:
    #    from_attributes = True
    model_config = ConfigDict(from_attributes=True)

File: frontend/lib/validation/request.ts
import { z } from "zod";

export const requestSchema = z.object({
  pickup_address: z
    .string()
    .trim()
    .min(1, "Pickup address is required")
    .max(255, "Pickup address is too long"),

  pickup_postcode: z
    .string()
    .trim()
    .min(1, "Pickup postcode is required")
    .max(12, "Pickup postcode is too long"),

  dropoff_address: z
    .string()
    .trim()
    .min(1, "Drop-off address is required")
    .max(255, "Drop-off address is too long"),

  dropoff_postcode: z
    .string()
    .trim()
    .min(1, "Drop-off postcode is required")
    .max(12, "Drop-off postcode is too long"),

  pickup_date: z
    .string()
    .min(1, "Pickup date is required"),

  pickup_time_slot: z
    .string()
    .min(1, "Pickup time slot is required"),

  helpers_needed: z.coerce
    .number()
    .int()
    .min(0, "Helpers cannot be negative"),

  pickup_floor: z.coerce
    .number()
    .int()
    .min(0, "Floor cannot be negative"),

  pickup_has_lift: z.boolean(),

  dropoff_floor: z.coerce
    .number()
    .int()
    .min(0, "Floor cannot be negative"),

  dropoff_has_lift: z.boolean(),

  pickup_loading_minutes: z.coerce
    .number()
    .int()
    .min(0, "Loading time cannot be negative"),

  dropoff_loading_minutes: z.coerce
    .number()
    .int()
    .min(0, "Loading time cannot be negative"),
});

export type RequestFormData = z.infer<typeof requestSchema>;